### PR TITLE
[Feature] Support Mixup and Cutmix for Recognizers.

### PR DESCRIPTION
**Pull request by irvingzhang0512**
Support Mixup and Cutmix for Recognizers.

## TODO
+ [x] codes of mixup/cutmix/register/recognizers
+ [x] unittest
+ [x] demo config for tsn
+ [x] changelog
+ [x] visualize input results(gif)
+ [x] ckpt

## Discussions
+ Since we are using soft-label, do we still need to calculate top1/top5 accuracy during training(not val)?
+ For now, we **CANNOT** inherit model settings and set `train_cfg`. A sample config is listed as follows. TypeError will be raised.

```python
_base_ = [
    '../../_base_/models/tsm_r50.py'
]

model = dict(
    train_cfg=dict(
        blending=dict(type="CutmixBlending", num_classes=400, alpha=.2)))
```

And error msg is `TypeError: train_cfg={'blending': {'type': 'CutmixBlending', 'num_classes': 400, 'alpha': 0.2}} in child config cannot inherit from base because train_cfg is a dict in the child config but is of type <class 'NoneType'> in base config. You may set `_delete_=True` to ignore the base config`

## visualize examples

mixup

mixup

cutmix
cutmix

## Results
+ model: tsm-r50 1x1x8 50e
+ dataset: sthv1

|configs|top1 acc(efficient/accuracy)|top5 acc(efficient/accuracy)|
|:-|:-:|:-:|
|mmaction2 model zoo|45.58 / 47.70|75.02 / 76.12|
|testing with model zoo ckpt|45.47 / 47.55|74.56 / 75.79|
|training with default config|45.82 / 47.90|74.38 / 76.02|
|mixup alpha=1.0|45.68 / 47.73|74.26 / 76.82|
|mixup alpha=0.2|46.35 / 48.49|75.07 / 76.88|
|cutmix alpha=0.2|45.92 / 47.46|75.23 / 76.71|

### CODE REVIEW

**Comment by irvingzhang0512** [comment]
@innerlee This pr is ready for review. Training models now.

**Comment by dreamerlin** [comment]
@congee524 Please help to check the config error

File: mmaction/datasets/__init__.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 from .audio_visual_dataset import AudioVisualDataset
 from .ava_dataset import AVADataset
 from .base import BaseDataset
+from .blending_utils import CutmixBlending, MixupBlending  # noqa
```
Comment (innerlee) [inline mmaction/datasets/__init__.py:7]: remove noqa

File: mmaction/datasets/blending_utils.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    def __call__(self, imgs, label, **kwargs):
+        """Blending data in a mini-batch.
+
+        Images are float tensors with the shape of (B, N, C, H, W) for 2D
+        recognizers or (B, N, C, T, H, W) for 3D recognizers.
+
+        Besides, labels are converted from hard labels to soft labels.
+        Hard labels are integer tensors with the shape of(B, 1) and all of the
+        elements are in the range [0, num_classes - 1].
+        Soft labels(probablity distribution over classes) are float tensors
```
Comment (innerlee) [inline mmaction/datasets/blending_utils.py:29]: space before brackets

File: mmaction/datasets/__init__.py (hunk 1 above)
Comment (dreamerlin) [reply mmaction/datasets/__init__.py:7]: And these should also be added in `__all__`

File: tests/test_data/test_blending.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import torch
+
+from mmaction.datasets.blending_utils import CutmixBlending, MixupBlending
```
Comment (dreamerlin) [inline tests/test_data/test_blending.py:3]: `from mmaction.datasets import CutmixBlending, MixupBlending` when `CutmixBlending, MixupBlending` are added in `__all__`

File: mmaction/datasets/blending_utils.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    This module is proposed in `mixup: Beyond Empirical Risk Minimization
+    <https://arxiv.org/abs/1710.09412>`_.
+    Code Reference https://github.com/open-mmlab/mmclassification/blob/master/mmcls/models/utils/mixup.py # noqa
+
+    Args:
+        num_classes (int): The number of classes.
+        alpha (float): Parameters for Beta distribution.
+    """
+
+    def __init__(self, num_classes, alpha=1.):
```
Comment (kennymckormick) [inline mmaction/datasets/blending_utils.py:71]: Better to use 0.2 as the default value of alpha, which is used in most scenarios

Comment (irvingzhang0512) [reply mmaction/datasets/blending_utils.py:71]: I'm not so sure about that. 

1. I follow VideoMix to set alpha to 1.0 for both mixup and cutmix.
2. After double checking the origin paper, for imagenet alpha should be in range [0.1, 0.4], but for cifar10/100 default value of alpha is 1.0.

I don't have enought resource to tune alpha right now. Maybe train more models in the future.

Comment (irvingzhang0512) [reply mmaction/datasets/blending_utils.py:71]: You are right after all, Orz...
For sthv1, alpha 0.2 is much better

**Comment by irvingzhang0512** [comment]
Since our kinetics400/sthv2 are somehow different, I train cutmix/mixup on sthv1.

sthv1 tsm-r50-1x1x8 results:

|configs|top1 acc(efficient/accuracy)|top5 acc(efficient/accuracy)|
|:-|:-:|:-:|
|mmaction2 model zoo|45.58 / 47.70|75.02 / 76.12|
|testing with model zoo ckpt|45.47 / 47.55|74.56 / 75.79|
|training with default config|45.82 / 47.90|74.38 / 76.02|
|mixup alpha=1.0|45.68 / 47.73|74.26 / 76.82|
|mixup alpha=0.2|46.35 / 48.49|75.07 / 76.88|
|cutmix alpha=0.2|45.92 / 47.46|75.23 / 76.71|

ckpt/json/log is here: https://wp.cethik.com:443/link/65C665F9E03EE42D9D479A70DACA53FA

Could anyone help to upload these?

**Comment by kennymckormick** [comment]
> Since our kinetics400/sthv2 are somehow different, I train cutmix/mixup on sthv1.
> 
> sthv1 tsm-r50-1x1x8 results:
> 
> configs	top1 acc(efficient/accuracy)	top5 acc(efficient/accuracy)
> mmaction2 model zoo	45.58 / 47.70	75.02 / 76.12
> testing with model zoo ckpt	45.47 / 47.55	74.56 / 75.79
> training with default config	45.82 / 47.90	74.38 / 76.02
> mixup alpha=1.0	45.68 / 47.73	74.26 / 76.82
> mixup alpha=0.2	46.35 / 48.49	75.07 / 76.88
> cutmix alpha=0.2	45.92 / 47.46	75.23 / 76.71
> ckpt/json/log is here: https://wp.cethik.com:443/link/65C665F9E03EE42D9D479A70DACA53FA
> 
> Could anyone help to upload these?

Sure

**Review by kennymckormick — APPROVED** [review]

**Comment by kennymckormick** [comment]
@innerlee  Please check this PR.

File: configs/recognition/tsm/tsm_r50_cutmix_1x1x8_50e_sthv1_rgb.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
+_base_ = [
+    '../../_base_/schedules/sgd_tsm_50e.py', '../../_base_/default_runtime.py'
+]
+
+# model settings
+# model settings# model settings
```
Comment (innerlee) [inline configs/recognition/tsm/tsm_r50_cutmix_1x1x8_50e_sthv1_rgb.py:6]: duplicate

File: configs/recognition/tsn/tsn_r50_video_mixup_1x1x8_100e_kinetics400_rgb.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    cls_head=dict(
+        type='TSNHead',
+        num_classes=400,
+        in_channels=2048,
+        spatial_type='avg',
+        consensus=dict(type='AvgConsensus', dim=1),
+        dropout_ratio=0.4,
+        init_std=0.01),
+    # model training and testing settings
+    # train_cfg=dict(
```
Comment (innerlee) [inline configs/recognition/tsn/tsn_r50_video_mixup_1x1x8_100e_kinetics400_rgb.py:22]: remove

**Comment by innerlee** [comment]
Thanks!

**Pull request merged**